zug/util.hpp

```diff
@@ -89,29 +90,87 @@ constexpr decltype(auto) invoke_composition(Fns&& fns, Args&&... args)
 }
 
 template <typename Fn, typename... Fns>
-struct composed
+struct composed : std::tuple<Fn, Fns...>
 {
-    std::tuple<Fn, Fns...> fns;
+    using base_t = std::tuple<Fn, Fns...>;
+
+    template <
+        typename TupleFns,
+        std::enable_if_t<
+            !std::is_same<composed, std::decay_t<TupleFns>>::value>* = nullptr>
+    composed(TupleFns&& fns)
+        : base_t{std::forward<TupleFns>(fns)}
+    {}
 
     template <typename... T>
     constexpr decltype(auto) operator()(T&&... xs) &
     {
-        return invoke_composition(fns, std::forward<T>(xs)...);
+        return invoke_composition(as_tuple(), std::forward<T>(xs)...);
     }
 
     template <typename... T>
     constexpr decltype(auto) operator()(T&&... xs) const&
     {
-        return invoke_composition(fns, std::forward<T>(xs)...);
+        return invoke_composition(as_tuple(), std::forward<T>(xs)...);
     }
 
     template <typename... T>
     constexpr decltype(auto) operator()(T&&... xs) &&
     {
-        return invoke_composition(std::move(fns), std::forward<T>(xs)...);
+        return invoke_composition(std::move(*this).as_tuple(),
+                                  std::forward<T>(xs)...);
+    }
+
+    template <typename... T>
+    base_t& as_tuple() &
+    {
+        return *this;
+    }
+
+    template <typename... T>
+    const base_t& as_tuple() const&
+    {
+        return *this;
+    }
+
+    template <typename... T>
+    base_t&& as_tuple() &&
+    {
+        return std::move(*this);
     }
 };
 
+template <typename T>
+constexpr bool is_composed_v = false;
+
+template <typename... Fns>
+constexpr bool is_composed_v<composed<Fns...>> = true;
+
+template <typename Composed,
+          std::enable_if_t<is_composed_v<Composed>>* = nullptr>
+decltype(auto) to_function_tuple(Composed&& c, meta::try_t)
+{
+    return std::forward<Composed>(c).as_tuple();
+}
+
+template <typename Fn>
+auto to_function_tuple(Fn&& fn, meta::catch_t)
+{
+    return std::make_tuple(std::forward<Fn>(fn));
+}
+
+template <typename T>
+auto to_function_tuple(T&& t)
```

Comment:
I think you want `decltype(auto)` here, also in the `catch_t` part of the implementation.

Reply:
Here it's missing, yes. But the `catch_t` doesn't require it as it's not dealing with references.